### DISCUSSION TITLE
How to deal with high order derivative that cannot be measured

### DISCUSSION THREAD

**Discussion by kouzhao** [do-mpc/do-mpc#328] [Q&A] [answered]
Hello,
I tried to use FOPDT model to control the temperature. Under this model, my input is heating power and the output is temperature. 
![图片](https://user-images.githubusercontent.com/20068083/203679870-e34f8ba6-4b5f-4c90-9f12-db47875046da.png)
I refer to your previous answer and use pade approximation to replace delay(Relaced tao with theta)
![图片](https://user-images.githubusercontent.com/20068083/203683231-0ce0a90e-b9ad-4fab-bfd7-5851bc59e5a6.png)
The time domain equation is(Replaced T1 with C*K):
![图片](https://user-images.githubusercontent.com/20068083/203680137-800c8020-a4d8-4571-b63e-34629723c8b5.png)

So I set the model according to the following codes:
`T = model.set_variable(var_type='_x', var_name='T')  #temprature`
`T_1 = model.set_variable(var_type='_x', var_name='T_1')  #First derivative of temperature`
`T_2 = model.set_variable(var_type='_x', var_name='T_2')  #Second derivative of temperature`
`P_heat = model.set_variable(var_type='_x', var_name='P_heat')  #Actual heating power`
`P_heat_1 = model.set_variable(var_type='_x', var_name='P_heat_1')  #First derivative of heating power`
`P_heat_2 = model.set_variable(var_type='_u', var_name='P_heat_2')  #Second derivative of heating power`

`T_set = model.set_variable(var_type='_tvp', var_name='T_set')`
`C = model.set_variable(var_type='_p', var_name='C')  #C=T/K`
`K = model.set_variable(var_type='_p', var_name='K')`
`T_E = model.set_variable(var_type='_p', var_name='T_E') #Environment temperature`
`theta = model.set_variable(var_type='_p', var_name='theta')`

`model.set_rhs('T', T_1)`
`model.set_rhs('T_1',T_2)`
`model.set_rhs('T_2', K*K*P_heat_2/C-6*K*K*P_heat_1/(C*theta)+12*K*K*P_heat/(C*theta*theta)-(6*C+theta*K)*T_2/(C*theta)-(6*theta*K+12*C)*T_1/(C*theta*theta)-12*K*(T-T_E)/(C*theta*theta))`
`model.set_rhs('P_heat', P_heat_1)`
`model.set_rhs('P_heat_1', P_heat_2)`
`model.set_rhs('err', T_set-T)`

Among the above variables, there are five state variables in total, but I can only measure `T` and `P_heat`, and the rest are derivatives of different orders,They perform well when working with the simulator,but if I want to transfer them to practical applications, how should I deal with variables that I cannot measure.

1. I'm confused about how to express the first or second derivative of a variables. I use set_rhs, this means that I have to set some values as state variables and measure them. However,in practice, I can only measure the temperature and control the heating power.

2. I doubt whether I have correctly understood the meaning of algebraic variables and state variables. I noticed that you used algebraic variables as the second derivative in the DIP example, and only provided the initial value of `theta` and `pos` during initialization. And in the subsequent simulation process, the simulator returned all state variables. Does this mean that we also need to measure these values in the actual scene?

BC

**Comment by 4flixt** [comment, marked as answer]
Hi @kouzhao,

you have stumbled upon some very common problems in control theory. Fortunately, there are solution for these problems.
1. You can convert any higher order ODE (say of degree n) into n ODEs of degree one. You will find books, youtube video etc. on this topic. 
2. If you have unmeasurable states, you will have to use an estimation algorithm. We call this an [observer](https://en.wikipedia.org/wiki/State_observer). do-mpc currently has a moving horizon estimator implemented for this purpose and we will soon add a Kalman filter and extended kalman filter. Again, there are many resources to learn about state-estimation (e.g. books, youtube, control lectures). 

Best,
4flixt
